src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1317 @@
+        ProjectId projectId, in CompilationOutputInfo info)
+    {
+        return ForkProject(
+            this.Solution.WithProjectCompilationOutputInfo(projectId, info),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectCompilationOutputInfo"/>
+    public SolutionCompilationState WithProjectDefaultNamespace(
+        ProjectId projectId, string? defaultNamespace)
+    {
+        return ForkProject(
+            this.Solution.WithProjectDefaultNamespace(projectId, defaultNamespace),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectChecksumAlgorithm"/>
+    public SolutionCompilationState WithProjectChecksumAlgorithm(
+        ProjectId projectId, SourceHashAlgorithm checksumAlgorithm)
+    {
+        return ForkProject(
+            this.Solution.WithProjectChecksumAlgorithm(projectId, checksumAlgorithm),
+            static stateChange => new CompilationAndGeneratorDriverTranslationAction.ReplaceAllSyntaxTreesAction(stateChange.NewProjectState, isParseOptionChange: false),
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectName"/>
+    public SolutionCompilationState WithProjectName(
+        ProjectId projectId, string name)
+    {
+        return ForkProject(
+            this.Solution.WithProjectName(projectId, name),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectFilePath"/>
+    public SolutionCompilationState WithProjectFilePath(
+        ProjectId projectId, string? filePath)
+    {
+        return ForkProject(
+            this.Solution.WithProjectFilePath(projectId, filePath),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectCompilationOptions"/>
+    public SolutionCompilationState WithProjectCompilationOptions(
+        ProjectId projectId, CompilationOptions options)
+    {
+        return ForkProject(
+            this.Solution.WithProjectCompilationOptions(projectId, options),
+            static stateChange => new CompilationAndGeneratorDriverTranslationAction.ProjectCompilationOptionsAction(stateChange.NewProjectState, isAnalyzerConfigChange: false),
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectParseOptions"/>
+    public SolutionCompilationState WithProjectParseOptions(
+        ProjectId projectId, ParseOptions options)
+    {
+        var stateChange = this.Solution.WithProjectParseOptions(projectId, options);
+
+        if (this.PartialSemanticsEnabled)
+        {
+            // don't fork tracker with queued action since access via partial semantics can become inconsistent (throw).
+            // Since changing options is rare event, it is okay to start compilation building from scratch.
```

Comment:
I've always hated this comment since it seemed like we were working around a limitation elsewhere. I think I get why it's here, but do we have benefits now of actually doing a replacement of the trees if there's a debug/release configuration switch and we can keep more things around?

Reply:
let's chat tomorrow.  I don't even think i understand the comment :)  happy to do more cleanup in a followup.  I wanted to keep things very safe here :)

Reply:
Yeah this is something to defer for later no matter what.